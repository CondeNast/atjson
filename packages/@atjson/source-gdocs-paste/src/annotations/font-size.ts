import { InlineAnnotation } from "@atjson/document";

export class FontSize extends InlineAnnotation<{ size: number }> {
  static vendorPrefix = "gdocs";
  static type = "ts_fs"; // Text style: font size
}
