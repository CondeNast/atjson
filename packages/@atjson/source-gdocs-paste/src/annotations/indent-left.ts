import { BlockAnnotation } from "@atjson/document";

export class IndentLeft extends BlockAnnotation<{ indent: number }> {
  static vendorPrefix = "gdocs";
  static type = "ps_il";
}
