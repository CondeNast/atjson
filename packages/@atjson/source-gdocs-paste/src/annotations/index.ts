export * from "./alignment";
export * from "./indent-left";
export * from "./bold";
export * from "./font-size";
export * from "./heading";
export * from "./horizontal-rule";
export * from "./italic";
export * from "./link";
export * from "./list-item";
export * from "./list";
export * from "./small-caps";
export * from "./strikethrough";
export * from "./underline";
export * from "./vertical-adjust";
