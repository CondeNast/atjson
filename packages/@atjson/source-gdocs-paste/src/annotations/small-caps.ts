import { InlineAnnotation } from "@atjson/document";

export class SmallCaps extends InlineAnnotation {
  static vendorPrefix = "gdocs";
  static type = "ts_sc"; // Text style: small caps
}
