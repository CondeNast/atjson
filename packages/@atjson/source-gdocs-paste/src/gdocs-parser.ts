import { AnnotationJSON } from "@atjson/document";

import { GDocsEntityMap, GDocsStyleSlice } from "./types";

import extractHorizontalRule from "./horizontal-rule";
import extractLinkStyles from "./link-styles";
import extractListStyles from "./list-styles";
import extractParagraphStyles from "./paragraph-styles";
import extractTextStyles from "./text-styles";

export interface GDocsPasteBuffer {
  [key: string]: any;
}

export interface Transforms {
  [key: string]: (
    styles: GDocsStyleSlice[],
    text: string,
    entityMap: GDocsEntityMap,
    trailing?: GDocsStyleSlice
  ) => AnnotationJSON[];
}

export default class GDocsParser {
  static transforms: Transforms = {
    text: extractTextStyles,
    paragraph: extractParagraphStyles,
    horizontal_rule: extractHorizontalRule,
    list: extractListStyles,
    link: extractLinkStyles,
  };

  gdocsSource: GDocsPasteBuffer;

  constructor(source: GDocsPasteBuffer) {
    this.gdocsSource = source;
  }

  getContent(): string {
    return this.gdocsSource.resolved.dsl_spacers;
  }

  getAnnotations(): AnnotationJSON[] {
    const styleSlices = this.gdocsSource.resolved.dsl_styleslices;
    const transforms = GDocsParser.transforms;
    const entityMap: GDocsEntityMap = this.gdocsSource.resolved.dsl_entitymap;

    let annotations: AnnotationJSON[] = [];
    for (let styleSlice of styleSlices) {
      let type: string = styleSlice.stsl_type;
      let styles: GDocsStyleSlice[] = styleSlice.stsl_styles;
      let trailing: GDocsStyleSlice = styleSlice.stsl_trailing;

      if (transforms[type]) {
        annotations.push(
          ...transforms[type](styles, this.getContent(), entityMap, trailing)
        );
      }
    }

    return annotations;
  }
}
