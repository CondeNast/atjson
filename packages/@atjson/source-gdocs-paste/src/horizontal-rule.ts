import { AnnotationJSON } from "@atjson/document";
import { GDocsStyleSlice } from "./types";

export default function extractHorizontalRule(
  styles: GDocsStyleSlice[],
): AnnotationJSON[] {
  let annotations: AnnotationJSON[] = [];

  for (let i = 0; i < styles.length; i++) {
    let style = styles[i];

    if (style === null) continue;

    annotations.push({
      type: "-gdocs-horizontal_rule",
      start: i,
      end: i + 1,
      attributes: {},
    });
  }

  return annotations;
}
