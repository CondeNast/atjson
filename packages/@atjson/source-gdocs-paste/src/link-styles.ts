import { AnnotationJSON } from "@atjson/document";
import { GDocsStyleSlice } from "./types";

export default function extractLinkStyles(
  linkStyles: GDocsStyleSlice[],
): AnnotationJSON[] {
  let currentLink: AnnotationJSON | null = null;
  let links: AnnotationJSON[] = [];

  for (let i = 0; i < linkStyles.length; i++) {
    let link = linkStyles[i];

    if (link === null) continue;

    // If we have a currentLink, and the current linkStyles[i] entry is not null,
    // then this means that the link is ending. Close up our current link and
    // push it into the list of found links.
    if (currentLink !== null) {
      currentLink.end = i;
      links.push(currentLink);

      currentLink = null;
    }

    // If the linkStyles[i] entry is not null, then we have a new link starting here.
    if (link.lnks_link !== null) {
      currentLink = {
        type: "-gdocs-lnks_link",
        start: i,
        end: -1,
        attributes: {
          "-gdocs-ulnk_url": link.lnks_link.ulnk_url,
          "-gdocs-lnk_type": link.lnks_link.lnk_type,
        },
      };
    }
  }

  // Add any unclosed links
  if (currentLink) {
    currentLink.end = linkStyles.length - 1;
    links.push(currentLink);
  }

  return links;
}
