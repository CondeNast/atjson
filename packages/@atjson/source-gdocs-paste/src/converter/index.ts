import {
  Annotation,
  BlockAnnotation,
  InlineAnnotation,
  ParseAnnotation,
  compareAnnotations,
  is,
} from "@atjson/document";
import OffsetSource, {
  Blockquote,
  Heading,
  LineBreak,
  ListItem,
  List,
  Paragraph,
  TextAlignment,
} from "@atjson/offset-annotations";
import GDocsSource from "../source";
import { Alignment, Heading as GDocsHeading } from "../annotations";
import { convertTables } from "./tables";
import { convertSmallCaps } from "./smallcaps";
import { convertDropCaps } from "./dropcaps";
import { IndentLeft } from "../annotations/indent-left";

// eslint-disable-next-line no-control-regex
const VERTICAL_TABS = /\u000B/g;
const NEWLINE_PARAGRAPH_SEPARATOR = /\n(\s*\n)*/g;

const ALIGNMENT = {
  0: undefined,
  1: TextAlignment.Center,
  2: TextAlignment.End,
  3: TextAlignment.Justify,
} as const;

GDocsSource.defineConverterTo(OffsetSource, (doc) => {
  // Remove all underlines that align with links, since
  // Google docs automatically does this when creating a link.
  // If necessary, underlined text can be added afterwards;
  // using this behavior ends up with unexpected consequences
  doc
    .where({ type: "-gdocs-ts_un" })
    .as("underline")
    .join(
      doc.where({ type: "-gdocs-lnks_link" }).as("link"),
      (underline, link) => underline.isAlignedWith(link)
    )
    .update(({ underline }) => {
      doc.removeAnnotation(underline);
    });

  doc.where({ type: "-gdocs-ts_bd" }).set({ type: "-offset-bold" });
  doc.where({ type: "-gdocs-ts_it" }).set({ type: "-offset-italic" });
  doc.where({ type: "-gdocs-ts_un" }).set({ type: "-offset-underline" });
  doc.where({ type: "-gdocs-ts_st" }).set({ type: "-offset-strikethrough" });
  doc.where({ type: "-gdocs-ts_sc" }).set({ type: "-offset-small-caps" });
  doc
    .where({ type: "-gdocs-ts_va", attributes: { "-gdocs-va": "sub" } })
    .set({ type: "-offset-subscript" })
    .unset("attributes.-gdocs-va");
  doc
    .where({ type: "-gdocs-ts_va", attributes: { "-gdocs-va": "sup" } })
    .set({ type: "-offset-superscript" })
    .unset("attributes.-gdocs-va");

  doc
    .where({ type: "-gdocs-horizontal_rule" })
    .set({ type: "-offset-horizontal-rule" });

  // Remove headings with level 100, 101, as these are Titles/Subtitles
  // in GDocs which is not yet supported, as these should be thought of
  // as non-body annotations
  doc.where({ type: "-gdocs-ps_hd" }).update((heading: GDocsHeading) => {
    let level = heading.attributes.level;
    if (level >= 100) {
      doc.removeAnnotation(heading);
    } else {
      doc.replaceAnnotation(
        heading,
        new Heading({
          id: heading.id,
          start: heading.start,
          end: heading.end,
          attributes: {
            level: level as 1 | 2 | 3 | 4 | 5 | 6,
            textAlignment: ALIGNMENT[heading.attributes.align],
          },
        })
      );
    }
  });

  // b_gt: 9 indicates an unordered list, but ordered lists have a variety of b_gt values
  doc
    .where({ type: "-gdocs-list", attributes: { "-gdocs-ls_b_gt": 9 } })
    .set({ type: "-offset-list", attributes: { "-offset-type": "bulleted" } });
  doc
    .where({ type: "-gdocs-list" })
    .set({ type: "-offset-list", attributes: { "-offset-type": "numbered" } });
  doc.where({ type: "-gdocs-list_item" }).set({ type: "-offset-list-item" });

  doc
    .where({ type: "-gdocs-lnks_link" })
    .set({ type: "-offset-link" })
    .rename({ attributes: { "-gdocs-ulnk_url": "-offset-url" } });

  let convertedLists = doc
    .where({ type: "-offset-list" })
    .as("list")
    .join(
      doc.where({ type: "-offset-list-item" }).as("listItems"),
      (list, listItem) =>
        list.start <= listItem.start && list.end >= listItem.end
    );

  convertedLists.update(({ list, listItems }) => {
    for (let newline of doc.match(
      NEWLINE_PARAGRAPH_SEPARATOR,
      list.start,
      list.end
    )) {
      let adjacentListItem = listItems.find(
        (listItem) => listItem.end === newline.start
      );
      if (adjacentListItem) {
        doc.addAnnotations(
          new ParseAnnotation({
            start: newline.start,
            end: newline.end,
            attributes: { reason: "list item separator" },
          })
        );
      }
    }
  });

  // Convert vertical tabs to line breaks
  for (let verticalTab of doc.match(VERTICAL_TABS)) {
    doc.addAnnotations(
      new LineBreak({
        start: verticalTab.start,
        end: verticalTab.end,
      }),
      new ParseAnnotation({
        start: verticalTab.start,
        end: verticalTab.end,
        attributes: {
          reason: "vertical tab",
        },
      })
    );
  }

  // Convert newlines to Paragraphs. Paragraphs must not cross the boundary of a BlockAnnotation, so
  // divide the document into 'block boundaries' and then look for single/multiple new lines within each
  // block boundary
  let blockBoundaries = doc
    .where((annotation: Annotation) => annotation instanceof BlockAnnotation)
    .reduce((boundaries: Set<number>, annotation: Annotation) => {
      boundaries.add(annotation.start);
      boundaries.add(annotation.end);
      return boundaries;
    }, new Set([0, doc.content.length]));

  [...blockBoundaries]
    .sort((boundary1, boundary2) => boundary1 - boundary2)
    .slice(1)
    .map((_, index, array) => {
      return {
        start: array[index - 1] || 0,
        end: array[index],
      };
    })
    .forEach(({ start, end }) => {
      // Multiple newlines indicate paragraph boundaries within the block boundary
      let paragraphBoundaries = doc.match(
        NEWLINE_PARAGRAPH_SEPARATOR,
        start,
        end
      );
      let lastEnd = start;
      for (let paragraphBoundary of paragraphBoundaries) {
        if (lastEnd < paragraphBoundary.start) {
          doc.addAnnotations(
            new Paragraph({
              start: lastEnd,
              end: paragraphBoundary.start,
            }),
            new ParseAnnotation({
              start: paragraphBoundary.start,
              end: paragraphBoundary.end,
              attributes: {
                reason: "paragraph boundary",
              },
            })
          );
        }
        lastEnd = paragraphBoundary.end;
      }

      // Close a remaining paragraph boundary at the block boundary
      if (start < lastEnd && lastEnd < end) {
        doc.addAnnotations(
          new Paragraph({
            start: lastEnd,
            end,
          })
        );
      }
    });

  doc.where({ type: "-gdocs-ps_al" }).update((align: Alignment) => {
    doc.replaceAnnotation(
      align,
      new Paragraph({
        start: align.start,
        end: align.end,
        attributes: {
          textAlignment: ALIGNMENT[align.attributes.align],
        },
      })
    );
  });

  // Interpreting paragraphs with a left indent as blockquotes
  doc.where({ type: "-gdocs-ps_il" }).update((bq) => {
    const text = doc.content.substring(bq.start, bq.end);
    const hasContent = bq.end > bq.start && text.trim().length > 0;

    console.log("ps_il annotation:", {
      start: bq.start,
      end: bq.end,
      length: bq.end - bq.start,
      text: text,
      indent: bq.attributes.indent,
      hasContent: hasContent,
    });

    // Only convert to blockquote if it has actual content
    if (hasContent) {
      doc.replaceAnnotation(
        bq,
        new Blockquote({
          start: bq.start,
          end: bq.end,
        })
      );
    } else {
      doc.removeAnnotation(bq);
    }
  });
  // GDocs can produce lists that have paragraphs in them that are
  // not wrapped by a list item. In these cases, split the list before
  // and after the paragraph, as necessary
  convertedLists
    .join(
      doc.where({ type: "-offset-paragraph" }).as("paragraphs"),
      ({ list }, paragraph) =>
        list.start <= paragraph.start && paragraph.end <= list.end
    )
    .update(({ list, listItems, paragraphs }) => {
      let blocks: Array<Paragraph | ListItem> = [...listItems, ...paragraphs];

      // Sort list items and paragraphs topologically so we can determine their
      // nested structure
      blocks.sort(compareAnnotations);

      let lastListItem = null;
      let currentList = list;
      let listEnd = list.end;
      for (let block of blocks) {
        // We have a list item, so mark it as the last seen.
        if (is(block, ListItem)) {
          lastListItem = block;

          // if the list was previously split, wrap the remainder in a new
          // list annotation. We only have to do this in this case when we've
          // found more list items that now lie outside the current list
          if (currentList.end < block.end) {
            currentList = new List({
              ...currentList,
              start: block.start,
              end: listEnd,
            });
            doc.addAnnotations(currentList);
          }
          continue;
        }

        // In this case we have a paragraph that is wrapped in a list item
        // We can just continue
        if (lastListItem && lastListItem.end >= block.end) {
          continue;
        }

        // We've found a paragraph with no wrapping list item. If the current
        // list overlaps this paragraph, truncate it to the beginning of the
        // of the paragraph.
        if (currentList.end > block.start) {
          currentList.end = block.start;
        }
      }
    });

  // Resolve single-item lists. New ones may have been created by splitting
  // lists
  doc
    .where({ type: "-offset-list" })
    .as("list")
    .join(
      doc.where({ type: "-offset-list-item" }).as("listItems"),
      (list, listItem) =>
        list.start === listItem.start && list.end === listItem.end
    )
    .update(({ list, listItems }) => {
      let insertionPoint = list.end;
      let item = listItems[0];
      doc.insertText(insertionPoint, "\uFFFC");
      doc.addAnnotations(
        new ParseAnnotation({
          start: insertionPoint,
          end: insertionPoint + 1,
          attributes: {
            reason: "object replacement character for single-item list",
          },
        })
      );
      item.end--;
    });

  convertTables(doc);

  // use font sizes to detect small caps and drop caps
  convertSmallCaps(doc);
  convertDropCaps(doc);

  // clean up font size annotations
  doc.where({ type: "-gdocs-ts_fs" }).remove();

  // adjust marks to not cover whitespace at the start / end positions
  doc
    .where((a) => a instanceof InlineAnnotation)
    .update((mark) => {
      let start = mark.start;
      let end = mark.end;
      while (
        start < end &&
        doc.content[start] &&
        doc.content[start].match(/\s/)
      ) {
        start++;
      }

      while (
        end > start &&
        doc.content[end - 1] &&
        doc.content[end - 1].match(/\s/)
      ) {
        end--;
      }

      mark.start = start;
      mark.end = end;
    });

  return doc;
});
