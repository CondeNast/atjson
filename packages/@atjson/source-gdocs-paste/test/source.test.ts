import * as fs from "fs";
import * as path from "path";
import { is } from "@atjson/document";
import { VerticalAdjust } from "../src/annotations";
import { GDocsPasteBuffer } from "../src/gdocs-parser";
import GDocsSource from "../src/index";

describe("@atjson/source-gdocs-paste", () => {
  describe("relatively complex document", () => {
    let pasteBuffer: GDocsPasteBuffer;

    beforeAll(() => {
      // https://docs.google.com/document/d/18pp4dAGx5II596HHGOLUXXcc6VKLAVRBUMLm9Ge8eOE/edit?usp=sharing
      let fixturePath = path.join(__dirname, "fixtures", "complex.json");
      pasteBuffer = JSON.parse(fs.readFileSync(fixturePath).toString());
    });

    it("has some json", () => {
      expect(pasteBuffer).toHaveProperty("resolved");
    });

    it("does not throw an error when instantiating with GDocsSource", () => {
      expect(GDocsSource.fromRaw(pasteBuffer)).toBeDefined();
    });

    it("correctly sets the content", () => {
      let gdocs = GDocsSource.fromRaw(pasteBuffer);
      expect(gdocs.content.length).toEqual(438);
      expect(gdocs.content).toMatchSnapshot();
    });

    it("extracts bold", () => {
      let gdocs = GDocsSource.fromRaw(pasteBuffer);
      let annotations = gdocs.where((a) => a.type === "ts_bd");
      expect(annotations.length).toEqual(2);

      let [a0, a1] = annotations;
      expect(gdocs.content.substring(a0.start, a0.end)).toEqual("simple te");
      expect(gdocs.content.substring(a1.start, a1.end)).toEqual("re is so");
    });

    it("extracts italic", () => {
      let gdocs = GDocsSource.fromRaw(pasteBuffer);
      let annotations = gdocs.where((a) => a.type === "ts_it");
      expect(annotations.length).toEqual(2);

      let [a0, a1] = [...annotations];
      expect(gdocs.content.substring(a0.start, a0.end)).toEqual("simple ");
      expect(gdocs.content.substring(a1.start, a1.end)).toEqual("some ");
    });

    it("extracts headings", () => {
      let gdocs = GDocsSource.fromRaw(pasteBuffer);
      let annotations = gdocs.where((a) => a.type === "ps_hd").sort();
      expect(annotations.length).toEqual(4);

      let [a0, a1, a2, a3] = [...annotations];

      expect(gdocs.content.substring(a0.start, a0.end)).toEqual("Heading 1");
      expect(a0.attributes.level).toEqual(1);

      expect(gdocs.content.substring(a1.start, a1.end)).toEqual("Heading 2");
      expect(a1.attributes.level).toEqual(2);

      expect(gdocs.content.substring(a2.start, a2.end)).toEqual("Title");
      expect(a2.attributes.level).toEqual(100);

      expect(gdocs.content.substring(a3.start, a3.end)).toEqual("Subtitle");
      expect(a3.attributes.level).toEqual(101);
    });

    describe("lists", () => {
      it("extracts lists", () => {
        let gdocs = GDocsSource.fromRaw(pasteBuffer);
        let annotations = gdocs.where((a) => a.type === "list");
        expect(annotations.length).toEqual(2);

        let [a0] = [...annotations];

        expect(gdocs.content.substring(a0.start, a0.end)).toEqual(
          "Here’s a numbered list\nAnd another item"
        );
        expect(a0.attributes.ls_id).toEqual("kix.trdi2u6o1bvt");
      });

      it("extracts list items", () => {
        let gdocs = GDocsSource.fromRaw(pasteBuffer);
        let annotations = gdocs.where((a) => a.type === "list_item");
        expect(annotations.length).toEqual(4);

        let [a0, a1] = [...annotations];

        expect(gdocs.content.substring(a0.start, a0.end)).toEqual(
          "Here’s a numbered list"
        );
        expect(a0.attributes.ls_id).toEqual("kix.trdi2u6o1bvt");
        expect(a0.attributes.ls_nest).toEqual(0);

        expect(gdocs.content.substring(a1.start, a1.end)).toEqual(
          "And another item"
        );
        expect(a1.attributes.ls_id).toEqual("kix.trdi2u6o1bvt");
        expect(a1.attributes.ls_nest).toEqual(0);
      });
    });

    it("extracts links", () => {
      let gdocs = GDocsSource.fromRaw(pasteBuffer);
      let annotations = gdocs.where((a) => a.type === "lnks_link");
      expect(annotations.length).toEqual(1);

      let [link] = [...annotations];

      expect(gdocs.content.substring(link.start, link.end)).toEqual(" is ");
      expect(link.attributes.ulnk_url).toEqual("https://www.google.com/");
      expect(link.attributes.lnk_type).toEqual(0);
    });
  });

  describe("a grab-bag of Google Docs features", () => {
    let gdocsBuffer: any;

    beforeAll(() => {
      // https://docs.google.com/document/d/18pp4dAGx5II596HHGOLUXXcc6VKLAVRBUMLm9Ge8eOE/edit?usp=sharing
      let fixturePath = path.join(
        __dirname,
        "fixtures",
        "formats-and-tabs.json"
      );
      gdocsBuffer = JSON.parse(fs.readFileSync(fixturePath).toString());
    });

    it("has some json", () => {
      expect(gdocsBuffer).toHaveProperty("resolved");
    });

    it("does not throw an error when instantiating with GDocsSource", () => {
      expect(GDocsSource.fromRaw(gdocsBuffer)).toBeDefined();
    });

    it("correctly sets the content", () => {
      let gdocs = GDocsSource.fromRaw(gdocsBuffer);
      expect(gdocs.content.length).toEqual(219);
      expect(gdocs.content).toMatchSnapshot();
    });

    it("extracts bold", () => {
      let gdocs = GDocsSource.fromRaw(gdocsBuffer);
      let annotations = gdocs.where((a) => a.type === "ts_bd");
      expect(annotations.length).toEqual(1);

      let [bold] = annotations;
      expect(gdocs.content.substring(bold.start, bold.end)).toEqual("bold");
    });

    it("extracts italic", () => {
      let gdocs = GDocsSource.fromRaw(gdocsBuffer);
      let annotations = gdocs.where((a) => a.type === "ts_it");
      expect(annotations.length).toEqual(1);

      let [italic] = annotations;
      expect(gdocs.content.substring(italic.start, italic.end)).toEqual(
        "italic"
      );
    });

    it("extracts underline", () => {
      let gdocs = GDocsSource.fromRaw(gdocsBuffer);
      let annotations = gdocs.where((a) => a.type === "ts_un");
      expect(annotations.length).toEqual(1);

      let [underline] = annotations;
      expect(gdocs.content.substring(underline.start, underline.end)).toEqual(
        "underlined"
      );
    });

    it("extracts horizontal rules", () => {
      let gdocs = GDocsSource.fromRaw(gdocsBuffer);
      let annotations = gdocs.where((a) => a.type === "horizontal_rule");
      expect(annotations.length).toEqual(1);

      let [hr] = annotations;
      expect(gdocs.content.substring(hr.start, hr.end)).toEqual("-");
    });

    it("extracts strikethrough", () => {
      let gdocs = GDocsSource.fromRaw(gdocsBuffer);
      let annotations = gdocs.where((a) => a.type === "ts_st");
      expect(annotations.length).toEqual(1);

      let [strikethrough] = annotations;
      expect(
        gdocs.content.substring(strikethrough.start, strikethrough.end)
      ).toEqual("strikethrough");
    });

    it("extracts vertical adjust", () => {
      let gdocs = GDocsSource.fromRaw(gdocsBuffer);
      let annotations = gdocs.where((a) => is(a, VerticalAdjust));
      expect(annotations.length).toEqual(2);

      let [superscript] = annotations.where(
        (annotation) => annotation.attributes.va === "sup"
      );
      let [subscript] = annotations.where(
        (annotation) => annotation.attributes.va === "sub"
      );
      expect(
        gdocs.content.substring(superscript.start, superscript.end)
      ).toEqual("TM");
      expect(gdocs.content.substring(subscript.start, subscript.end)).toEqual(
        "2"
      );
    });
  });

  describe("list styles", () => {
    let gdocsBuffer: any;

    beforeAll(() => {
      let fixturePath = path.join(__dirname, "fixtures", "list-styles.json");
      gdocsBuffer = JSON.parse(fs.readFileSync(fixturePath).toString());
    });

    it("creates the right number of list annotations", () => {
      let gdocs = GDocsSource.fromRaw(gdocsBuffer);
      let lists = gdocs.where((a) => a.type === "list");

      expect(lists.length).toEqual(2);
    });

    it("captures list-specific attributes", () => {
      let gdocs = GDocsSource.fromRaw(gdocsBuffer);
      let lists = gdocs.where((a) => a.type === "list");
      let expectedShape = expect.objectContaining({
        ls_b_gs: expect.anything(),
        ls_b_gt: expect.anything(),
        ls_b_a: expect.anything(),
      });

      for (let list of lists) {
        expect(list.attributes).toEqual(expectedShape);
      }
    });

    it("distinguishes numbered from bulleted lists", () => {
      let gdocs = GDocsSource.fromRaw(gdocsBuffer);
      let lists = gdocs.annotations
        .filter((a) => a.type === "list")
        .filter((a) => a.attributes.ls_b_gt === 9);

      expect(lists.length).toEqual(1);
    });
  });

  describe("partial link pastes", () => {
    let gdocsBuffer: any;

    beforeAll(() => {
      let fixturePath = path.join(__dirname, "fixtures", "partial.json");
      gdocsBuffer = JSON.parse(fs.readFileSync(fixturePath).toString());
    });

    it("creates the right number of link annotations", () => {
      let gdocs = GDocsSource.fromRaw(gdocsBuffer);
      let links = gdocs.annotations.filter((a) => a.type === "lnks_link");

      expect(links.length).toEqual(1);
    });
  });

  describe("partial list pastes", () => {
    let pasteBuffer: any;

    beforeAll(() => {
      // https://docs.google.com/document/d/1PKNoasDTf0Pj71vJs4MAi9zOrWDA3TDSNj0RFXWoCp4/edit
      let fixturePath = path.join(
        __dirname,
        "fixtures",
        "list-styles-partial.json"
      );
      pasteBuffer = JSON.parse(fs.readFileSync(fixturePath).toString());
    });

    it("creates the right number of list and list-item annotations", () => {
      let gdocs = GDocsSource.fromRaw(pasteBuffer);
      let listAndItems = gdocs
        .where((a) => a.type === "list")
        .as("list")
        .join(
          gdocs.where((a) => a.type === "list_item").as("listItems"),
          (l, r) => l.start <= r.start && l.end >= r.end
        );

      expect(listAndItems.toJSON()).toMatchObject([
        {
          list: { start: 0, end: 22, type: "-gdocs-list" },
          listItems: [
            { start: 0, end: 8, type: "-gdocs-list_item" },
            { start: 9, end: 22, type: "-gdocs-list_item" },
          ],
        },
        {
          list: { start: 41, end: 68, type: "-gdocs-list" },
          listItems: [
            { start: 41, end: 54, type: "-gdocs-list_item" },
            { start: 55, end: 68, type: "-gdocs-list_item" },
          ],
        },
        {
          list: { start: 89, end: 102, type: "-gdocs-list" },
          listItems: [{ start: 89, end: 102, type: "-gdocs-list_item" }],
        },
      ]);
    });
  });
});
