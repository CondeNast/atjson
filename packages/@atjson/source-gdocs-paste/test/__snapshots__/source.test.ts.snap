// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@atjson/source-gdocs-paste a grab-bag of Google Docs features correctly sets the content 1`] = `
"	This paragraph is indented! Here is some underlined text and some italic text and some bold text. This is superscriptTM and some subscript O2 and some strikethrough text. This is a link. This text has a comment.
	
-


"
`;

exports[`@atjson/source-gdocs-paste relatively complex document correctly sets the content 1`] = `
"This is a simple test.
*
Here is some text

Hereâ€™s a numbered list
And another item

Hereâ€™s a bulleted list
With two items, like before

Here is a poem
  With some text ahead
    Four spaces this time and a suggestion
Now no spaces
Hereâ€™s a line with some different line spacing (1.5)
And hereâ€™s a double spaced line
Heading 1
Heading 2
Title
Subtitle
UPPER CASE
Title Case is fun
This is superscript!
This is subscript
abdsfdf


"
`;
